src/margo-efirst-pool.c

```diff
@@ -0,0 +1,333 @@
+} pool_t;
+
+static ABT_unit_type pool_unit_get_type(ABT_unit unit)
+{
+    unit_t* p_unit = (unit_t*)unit;
+    if (p_unit->flag & IS_THREAD) {
+        return ABT_UNIT_TYPE_THREAD;
+    } else {
+        return ABT_UNIT_TYPE_TASK;
+    }
+}
+
+static ABT_thread pool_unit_get_thread(ABT_unit unit)
+{
+    unit_t* p_unit = (unit_t*)unit;
+    return p_unit->thread;
+}
+
+static ABT_task pool_unit_get_task(ABT_unit unit)
+{
+    unit_t* p_unit = (unit_t*)unit;
+    return p_unit->task;
+}
+
+static ABT_bool pool_unit_is_in_pool(ABT_unit unit)
+{
+    unit_t* p_unit = (unit_t*)unit;
+    return p_unit->flag & IS_IN_POOL;
+}
+
+static ABT_unit pool_unit_create_from_thread(ABT_thread thread)
+{
+    unit_t* p_unit = (unit_t*)calloc(1, sizeof(unit_t));
+    p_unit->thread = thread;
+    p_unit->flag   = IS_THREAD;
+    return (ABT_unit)p_unit;
+}
+
+static ABT_unit pool_unit_create_from_task(ABT_task task)
+{
+    unit_t* p_unit = (unit_t*)calloc(1, sizeof(unit_t));
+    p_unit->task   = task;
+    return (ABT_unit)p_unit;
+}
+
+static void pool_unit_free(ABT_unit* p_unit)
+{
+    free(*p_unit);
+    *p_unit = ABT_UNIT_NULL;
+}
+
+static int pool_init(ABT_pool pool, ABT_pool_config config)
+{
+    (void)config;
+    pool_t* p_pool = (pool_t*)calloc(1, sizeof(pool_t));
+    p_pool->queue  = create_queue(32);
+    pthread_mutex_init(&p_pool->mutex, NULL);
+    pthread_cond_init(&p_pool->cond, NULL);
+    ABT_pool_set_data(pool, (void*)p_pool);
+    return ABT_SUCCESS;
+}
+
+static size_t pool_get_size(ABT_pool pool)
+{
+    pool_t* p_pool;
+    ABT_pool_get_data(pool, (void**)&p_pool);
+    size_t prio_size = p_pool->queue->prio.size;
+    size_t fifo_size = p_pool->queue->fifo.size;
+    return prio_size + fifo_size;
+}
+
+static void pool_push(ABT_pool pool, ABT_unit unit)
+{
+    pool_t* p_pool;
+    ABT_pool_get_data(pool, (void**)&p_pool);
+    unit_t* p_unit = (unit_t*)unit;
+    pthread_mutex_lock(&p_pool->mutex);
+    if (p_unit->priority == 0) {
+        p_pool->num += 1;
+        p_unit->priority = p_pool->num;
+    }
+    queue_push(p_pool->queue, p_unit);
+    pthread_cond_signal(&p_pool->cond);
+    pthread_mutex_unlock(&p_pool->mutex);
+}
+
+static ABT_unit pool_pop(ABT_pool pool)
+{
+    pool_t* p_pool;
+    ABT_pool_get_data(pool, (void**)&p_pool);
+    pthread_mutex_lock(&p_pool->mutex);
+    unit_t* p_unit = queue_pop(p_pool->queue);
+    pthread_mutex_unlock(&p_pool->mutex);
+    return p_unit ? (ABT_unit)p_unit : ABT_UNIT_NULL;
+}
+
+static inline void convert_double_sec_to_timespec(struct timespec* ts_out,
+                                                  double           seconds)
+{
+    ts_out->tv_sec  = (time_t)seconds;
+    ts_out->tv_nsec = (long)((seconds - ts_out->tv_sec) * 1000000000.0);
+}
+
+static ABT_unit pool_pop_timedwait(ABT_pool pool, double abstime_secs)
+{
+    pool_t* p_pool;
+    ABT_pool_get_data(pool, (void**)&p_pool);
+    pthread_mutex_lock(&p_pool->mutex);
+    if (p_pool->queue->prio.size == 0) {
```

Comment:
I have a question here - should this also account for `fifo.size` so that it doesn't block if there is work ready in that queue?

Reply:
Thanks! Fixed.